bundles/core/org.eclipse.smarthome.core/src/main/java/org/eclipse/smarthome/core/common/osgi/ResourceBundleClassLoader.java

```diff
@@ -172,15 +187,26 @@ private Charset getResourceCharset(URL url) {
         return charset;
     }
 
+    /**
+     * The method tests whether the file content can be decoded with the given character set.
+     *
+     * @param f
+     *            name of the file that will be tested
+     * @param charset
+     *            character set to which the file is tested
+     * @return
+     *         passed character set if the file could be decoded with this character set or null if the decode fails
+     */
     private Charset detectCharset(String f, Charset charset) {
         try {
             InputStream in = super.getResourceAsStream(f);
             BufferedInputStream input = new BufferedInputStream(in);
             CharsetDecoder decoder = charset.newDecoder();
             decoder.reset();
             byte[] buffer = new byte[512];
-            boolean identified = false;
-            while ((input.read(buffer) != -1) && (!identified)) {
```

Comment:
> The downside of this implementation is that it potentially consumes more memory as it does not process the stream in chunks, but loads it completely into memory before starting to process it. As we are talking about text files, I would assume that this is acceptable, though.

I had the same thoughts. But I don't see any other reliable method. But I can optimize the code and create the the byte array only once per file by moving the code out the loop.

>One very last question: Are we sure that the stream returned by getResourceAsStream(f) is always correctly closed?

The documentation is not clear but I found some hints that it would be better to close it. I will add a close method and test it.